### PR TITLE
:bug: Wait until the resources is ready to start controller-manager

### PR DESCRIPTION
**Pull request by yanmxa**
## Summary

The controller runtime manager crashes if it watches a crd resource that hasn't existed since version 0.15.  So if we want to run the global hub operator under a bare kube environment, it needs to detect the necessary resources to be ready to start watching them. 

### Solution

1. We will first start a crd controller to detect the necessary resources, like ACM or Kafka. 
2. To avoid the operator memory being overused. Instead of watching CRD directly, let the  controller use `WatchesMetadata`
3. Once the ACM resources are ready, then the crd controller starts the addon, and global hub controllers
4. The crd controller also records the states of Kafka resources and the triggered addon and global hub controllers.

## Related issue(s)
References: https://github.com/kubernetes-sigs/controller-runtime/issues/2456 
Fixes # https://issues.redhat.com/browse/ACM-10955

### CODE REVIEW

File: operator/main.go [hunk 1]
```diff
@@ -136,6 +139,11 @@ func doMain(ctx context.Context, cfg *rest.Config) int {
 	operatorConfig := parseFlags()
 	utils.PrintVersion(setupLog)
 
+	if err := ResourceReady(ctx, cfg); err != nil {
```
Comment (clyang82) [inline operator/main.go:142]: in this case, the operator cannot be started without those resource. is it correct? if so, the CVP test cannot be passed with this fix.

File: operator/main.go [hunk 2]
```diff
@@ -296,107 +218,100 @@ func getManager(restConfig *rest.Config, electionConfig *commonobjects.LeaderEle
...
+
+		// addon installer, global hub controller
 		&clusterv1.ManagedCluster{}: {
 			Label: labels.SelectorFromSet(labels.Set{"vendor": "OpenShift"}),
 		},
-		&workv1.ManifestWork{}: {
-			Label: labelSelector,
-		},
+		// addon installer, global hub controller
 		&addonv1alpha1.ClusterManagementAddOn{}: {
```
Comment (clyang82) [inline operator/main.go:300]: `ClusterManagementAddOn` is ACM CRD. the operator cannot be started without it.

File: operator/pkg/controllers/crd/crd_controller.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+	addonController                   *addon.AddonController
+	globalHubConditionControllerReady bool
+	globalHubControllerReady          bool
+	backupControllerReady             bool
+}
+
+func (r *CrdController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.resources[req.Name] = true
+	if !r.readyToWatchACMResources() {
+		return ctrl.Result{}, nil
```
Comment (clyang82) [inline operator/pkg/controllers/crd/crd_controller.go:70]: How can we dynamically update cache when those resources are ready?

**Review by clyang82 — CHANGES_REQUESTED** [review]
Did you test the operator in pure OpenShift cluster?

File: operator/main.go (hunk 2 above)
Comment (yanmxa) [reply operator/main.go:300]: The `ClusterManagementAddOn` can be added to the cache when starting the manager. It only crashes when `Watch` a CRD resource that doesn't exist in the cluster.

File: operator/pkg/controllers/crd/crd_controller.go (hunk 3 above)
Comment (yanmxa) [reply operator/pkg/controllers/crd/crd_controller.go:70]: It provides a lower method to update the cache when dynamically adding the controller to the manager, like: 
```go
.WatchesRawSource(source.Kind(installerCache, &clusterv1.ManagedCluster{}),
	&handler.EnqueueRequestForObject{}, builder.WithPredicates(clusterPred)).
```
However, the cache isn't the reason to break down the operator. Do we need to replace the manager cache with the lower interface?

**Comment by yanmxa** [comment]
If the crd resources aren't ready. the operator will be running with the following log:
```bash
oc logs -f -l name=multicluster-global-hub-operator -n multicluster-global-hub
2024-04-18T05:25:41.385Z        INFO    setup   Go OS/Arch: linux/amd64
2024-04-18T05:25:41.402Z        INFO    setup   starting manager
2024-04-18T05:25:41.402Z        INFO    controller-runtime.metrics      Starting metrics server
2024-04-18T05:25:41.402Z        INFO    starting server {"kind": "health probe", "addr": "[::]:8081"}
2024-04-18T05:25:41.402Z        INFO    controller-runtime.metrics      Serving metrics server  {"bindAddress": ":8080", "secure": false}
I0418 05:25:41.402653       1 leaderelection.go:250] attempting to acquire leader lease multicluster-global-hub/multicluster-global-hub-operator-lock...
I0418 05:25:41.407270       1 leaderelection.go:260] successfully acquired lease multicluster-global-hub/multicluster-global-hub-operator-lock
2024-04-18T05:25:41.408Z        INFO    Starting EventSource    {"controller": "CRDController", "source": "kind source: *v1.PartialObjectMetadata"}
2024-04-18T05:25:41.408Z        INFO    Starting Controller     {"controller": "CRDController"}
2024-04-18T05:25:41.510Z        INFO    Starting workers        {"controller": "CRDController", "worker count": 1}
```

**Comment by yanmxa** [comment]
> Did you test the operator in pure OpenShift cluster?

Yes. It works now.

File: operator/pkg/controllers/crd/crd_controller.go (hunk 3 above)
Comment (yanmxa) [reply operator/pkg/controllers/crd/crd_controller.go:70]: I try to build a dynamic cache but It cannot work.
```bash
func ACMCache(mgr ctrl.Manager) (cache.Cache, error) {
	cacheOptions := cache.Options{
		Scheme: mgr.GetScheme(),
		Mapper: mgr.GetRESTMapper(),
	}
	cacheOptions.ByObject = map[client.Object]cache.ByObject{
		// addon installer, global hub controller
		&clusterv1.ManagedCluster{}: {
			Label: labels.SelectorFromSet(labels.Set{"vendor": "OpenShift"}),
		},
		// addon installer, global hub controller
		&addonv1alpha1.ClusterManagementAddOn{}: {
			Label: labelSelector,
		},
		// addon installer
		&addonv1alpha1.ManagedClusterAddOn{}: {
			Label: labelSelector,
		},
		// global hub controller
		&promv1.ServiceMonitor{}: {
			Label: labelSelector,
		},
		// global hub controller
		&subv1alpha1.Subscription{}: {},
	}
	extendCache, err := cache.New(mgr.GetConfig(), cacheOptions)
	if err != nil {
		return nil, err
	}
	return extendCache, err
}
```

**Comment by yanmxa** [comment]
override "ci/prow/sonarcloud"

**Comment by yanmxa** [comment]
When start the operator in a OCP without ACM: 
```bash
root@myan workspace/hub-of-hubs (br_wait_ready) » oc logs -f -l name=multicluster-global-hub-operator -n multicluster-global-hub
2024-04-26T16:01:56.522Z        INFO    setup   Go OS/Arch: linux/amd64
2024-04-26T16:01:56.540Z        INFO    setup   starting manager
2024-04-26T16:01:56.540Z        INFO    controller-runtime.metrics      Starting metrics server
2024-04-26T16:01:56.540Z        INFO    starting server {"kind": "health probe", "addr": "[::]:8081"}
2024-04-26T16:01:56.540Z        INFO    controller-runtime.metrics      Serving metrics server  {"bindAddress": ":8080", "secure": false}
I0426 16:01:56.540571       1 leaderelection.go:250] attempting to acquire leader lease multicluster-global-hub/multicluster-global-hub-operator-lock...
I0426 16:01:56.545371       1 leaderelection.go:260] successfully acquired lease multicluster-global-hub/multicluster-global-hub-operator-lock
2024-04-26T16:01:56.545Z        INFO    Starting EventSource    {"controller": "CRDController", "source": "kind source: *v1.PartialObjectMetadata"}
2024-04-26T16:01:56.545Z        INFO    Starting Controller     {"controller": "CRDController"}
2024-04-26T16:01:56.646Z        INFO    Starting workers        {"controller": "CRDController", "worker count": 1}
```

**Comment by yanmxa** [comment]
After installing ACM. It will start all the other controllers:
```bash
I0426 16:07:29.302458       1 utils.go:241] Wait MulticlusterGlobalHub created
2024-04-26T16:07:29.302Z        INFO    Starting EventSource    {"controller": "addonInstaller", "source": "kind source: *v1.ManagedCluster"}
2024-04-26T16:07:29.302Z        INFO    Starting EventSource    {"controller": "addonInstaller", "source": "kind source: *v1alpha1.ManagedClusterAddOn"}
2024-04-26T16:07:29.302Z        INFO    Starting EventSource    {"controller": "addonInstaller", "source": "kind source: *v1alpha1.ClusterManagementAddOn"}
2024-04-26T16:07:29.302Z        INFO    Starting EventSource    {"controller": "addonInstaller", "source": "kind source: *v1.Secret"}
2024-04-26T16:07:29.302Z        INFO    Starting Controller     {"controller": "addonInstaller"}
2024-04-26T16:07:29.303Z        INFO    Starting Controller     {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub"}
2024-04-26T16:07:29.303Z        INFO    Starting workers        {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "worker count": 1}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "conditionController", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1alpha4.MulticlusterGlobalHub"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "conditionController", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.Deployment"}
2024-04-26T16:07:29.303Z        INFO    Starting Controller     {"controller": "conditionController", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1alpha4.MulticlusterGlobalHub"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.Deployment"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.StatefulSet"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.Service"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.ServiceAccount"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.Secret"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.Role"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.RoleBinding"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.Route"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.MutatingWebhookConfiguration"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.ConfigMap"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.Namespace"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.ClusterRole"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.ClusterRoleBinding"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.Secret"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1alpha1.ClusterManagementAddOn"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.ManagedCluster"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.ServiceMonitor"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "multiclusterglobalhub", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1alpha1.Subscription"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "backupController", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1alpha4.MulticlusterGlobalHub"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "backupController", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.Secret"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "backupController", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.ConfigMap"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "backupController", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.PersistentVolumeClaim"}
2024-04-26T16:07:29.303Z        INFO    Starting EventSource    {"controller": "backupController", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "source": "kind source: *v1.MultiClusterHub"}
2024-04-26T16:07:29.303Z        INFO    Starting Controller     {"controller": "backupController", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub"}
2024-04-26T16:07:29.534Z        INFO    Starting workers        {"controller": "addonInstaller", "worker count": 1}
2024-04-26T16:07:29.534Z        INFO    Starting workers        {"controller": "backupController", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "worker count": 1}
2024-04-26T16:07:29.535Z        INFO    Starting workers        {"controller": "conditionController", "controllerGroup": "operator.open-cluster-management.io", "controllerKind": "MulticlusterGlobalHub", "worker count": 1}
```

**Review by clyang82 — APPROVED** [review]
Perfect

**Pull request merged**